File: openry/orchestrator/validation.py
"""Hard-code validation engine for sub_step results."""

from __future__ import annotations

import json
import os
import re
import subprocess
from typing import Any

from openry.db import (
    get_conn,
    get_task_state,
    insert_validation_result,
)


def validate_step(run_id: str, step_config: dict[str, Any]) -> tuple[bool, str]:
    """Run all validation rules against a completed sub_step.

    Args:
        run_id: The run to validate
        step_config: The sub_step YAML config (including validation rules)

    Returns:
        (passed, failure_reason) — failure_reason is "" on success
    """
    state = get_task_state(run_id)
    if not state:
        return False, f"run_id not found: {run_id}"

    try:
        payload = json.loads(state.get("payload", "{}"))
    except (json.JSONDecodeError, TypeError):
        payload = {}

    # 1. expect_payload check
    if step_config.get("expect_payload") and not payload:
        _log_validation(run_id, "expect_payload", None, False, "no payload provided")
        return False, "expect_payload=True but no payload provided"

    # 2. payload_keys check (implicit validation)
    for key in step_config.get("payload_keys", []):
        if key not in payload:
            _log_validation(run_id, "payload_has_key", key, False, f"missing key: {key}")
            return False, f"missing required payload key: {key}"

    # 3. Explicit validation rules
    for rule in step_config.get("validation", []):
        rule_type = rule.get("type", "")
        rule_params = json.dumps({k: v for k, v in rule.items() if k != "type"})
        passed = True
        reason = ""

        if rule_type == "payload_has_key":
            passed = rule["key"] in payload
            reason = "" if passed else f"missing key: {rule['key']}"

        elif rule_type == "payload_value_matches":
            value = str(payload.get(rule["key"], ""))
            passed = bool(re.match(rule["regex"], value))
            reason = "" if passed else f"value mismatch: {rule['key']}={value}"

        elif rule_type == "payload_values_equal":
            passed = payload.get(rule["key_a"]) == payload.get(rule["key_b"])
            reason = "" if passed else f"values not equal: {rule['key_a']} != {rule['key_b']}"

        elif rule_type == "file_exists":
            path = rule["path"]
            passed = os.path.exists(path)
            reason = "" if passed else f"file not found: {path}"

        elif rule_type == "file_contains":
            path = rule["path"]
            if not os.path.exists(path):
                passed = False
                reason = f"file not found: {path}"
            else:
                with open(path) as f:
                    passed = rule["contains"] in f.read()
                reason = "" if passed else f"missing content: {rule['contains']}"

        elif rule_type == "command":
            result = subprocess.run(
                rule["run"], shell=True, capture_output=True, text=True, timeout=60,
            )
            passed = result.returncode == 0
            reason = f"exit_code={result.returncode}" if not passed else ""

        elif rule_type == "command_output_contains":
            result = subprocess.run(
                rule["run"], shell=True, capture_output=True, text=True, timeout=60,
            )
            passed = rule["contains"] in result.stdout
            reason = "" if passed else f"output missing: {rule['contains']}"

        elif rule_type == "db_query":
            conn = get_conn()
            row = conn.execute(rule["query"]).fetchone()
            conn.close()
            passed = row is not None
            reason = "" if passed else "query returned no rows"

        else:
            passed = False
            reason = f"unknown rule type: {rule_type}"

        _log_validation(run_id, rule_type, rule_params, passed, reason)
        if not passed:
            return False, reason

    return True, ""


def _log_validation(
    run_id: str,
    rule_type: str,
    rule_params: str | None,
    passed: bool,
    message: str,
) -> None:
    """Log a single validation result to DB."""
    try:
        insert_validation_result(run_id, rule_type, rule_params, passed, message)
    except Exception:
        pass  # Validation logging is best-effort

File: orchestrator-plugin/src/orchestrator/validation.ts
/**
 * 硬代码验证引擎 — 对 completed 的 sub_step 执行验证规则。
 *
 * 与 Python openry/orchestrator/validation.py 逻辑完全一致。
 */
import * as fs from "node:fs";
import { execSync } from "node:child_process";
import Database from "better-sqlite3";

export type ValidationRule =
  | { type: "payload_has_key"; key: string }
  | { type: "payload_value_matches"; key: string; regex: string }
  | { type: "payload_values_equal"; key_a: string; key_b: string }
  | { type: "file_exists"; path: string }
  | { type: "file_contains"; path: string; contains: string }
  | { type: "command"; run: string }
  | { type: "command_output_contains"; run: string; contains: string }
  | { type: "db_query"; query: string };

export type StepConfig = {
  expect_payload?: boolean;
  payload_keys?: string[];
  validation?: ValidationRule[];
  on_validation_fail?: string;
};

export function validateStep(
  db: Database.Database,
  runId: string,
  stepConfig: StepConfig,
): { passed: boolean; reason: string } {
  const state = db
    .prepare("SELECT payload FROM task_state WHERE run_id = ?")
    .get(runId) as { payload: string } | undefined;

  let payload: Record<string, unknown> = {};
  if (state) {
    try {
      payload = JSON.parse(state.payload);
    } catch {
      // keep empty
    }
  }

  // 1. expect_payload check
  if (stepConfig.expect_payload && Object.keys(payload).length === 0) {
    logValidation(db, runId, "expect_payload", null, false, "no payload provided");
    return { passed: false, reason: "expect_payload=True but no payload provided" };
  }

  // 2. payload_keys check (implicit validation)
  for (const key of stepConfig.payload_keys ?? []) {
    if (!(key in payload)) {
      logValidation(db, runId, "payload_has_key", key, false, `missing key: ${key}`);
      return { passed: false, reason: `missing required payload key: ${key}` };
    }
  }

  // 3. Explicit validation rules
  for (const rule of stepConfig.validation ?? []) {
    const ruleParams = JSON.stringify(
      Object.fromEntries(Object.entries(rule).filter(([k]) => k !== "type")),
    );
    let passed = true;
    let reason = "";

    switch (rule.type) {
      case "payload_has_key":
        passed = rule.key in payload;
        reason = passed ? "" : `missing key: ${rule.key}`;
        break;

      case "payload_value_matches": {
        const val = String(payload[rule.key] ?? "");
        passed = new RegExp(rule.regex).test(val);
        reason = passed ? "" : `value mismatch: ${rule.key}=${val}`;
        break;
      }

      case "payload_values_equal":
        passed = payload[rule.key_a] === payload[rule.key_b];
        reason = passed
          ? ""
          : `values not equal: ${rule.key_a} != ${rule.key_b}`;
        break;

      case "file_exists":
        passed = fs.existsSync(rule.path);
        reason = passed ? "" : `file not found: ${rule.path}`;
        break;

      case "file_contains":
        if (!fs.existsSync(rule.path)) {
          passed = false;
          reason = `file not found: ${rule.path}`;
        } else {
          const content = fs.readFileSync(rule.path, "utf-8");
          passed = content.includes(rule.contains);
          reason = passed ? "" : `missing content: ${rule.contains}`;
        }
        break;

      case "command": {
        try {
          execSync(rule.run, { timeout: 60_000, encoding: "utf-8" });
          passed = true;
        } catch (err: unknown) {
          passed = false;
          const code =
            err && typeof err === "object" && "status" in err
              ? (err as { status: number }).status
              : "error";
          reason = `exit_code=${code}`;
        }
        break;
      }

      case "command_output_contains": {
        try {
          const output = execSync(rule.run, {
            timeout: 60_000,
            encoding: "utf-8",
          });
          passed = output.includes(rule.contains);
          reason = passed ? "" : `output missing: ${rule.contains}`;
        } catch {
          passed = false;
          reason = "command failed";
        }
        break;
      }

      case "db_query": {
        try {
          const row = db.prepare(rule.query).get();
          passed = row !== undefined;
          reason = passed ? "" : "db query returned no rows";
        } catch {
          passed = false;
          reason = "db query failed";
        }
        break;
      }

      default:
        passed = true;
    }

    logValidation(db, runId, rule.type, ruleParams, passed, reason);

    if (!passed) {
      return { passed: false, reason };
    }
  }

  return { passed: true, reason: "" };
}

function logValidation(
  db: Database.Database,
  runId: string,
  ruleType: string,
  ruleParams: string | null,
  passed: boolean,
  message: string,
): void {
  db.prepare(
    `INSERT INTO validation_results (run_id, rule_type, rule_params, passed, message)
     VALUES (?, ?, ?, ?, ?)`,
  ).run(runId, ruleType, ruleParams, passed ? 1 : 0, message);
}
